core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -1065,6 +1067,157 @@ private void reduceCasts(RexCall outerCast) {
       }
     }
   }
+
+  /** Pushes a condition into an OR if they reference only 1 variable
+   *
+   * Example: a=1 && (a=1 || a=2)
+   */
+  protected static class PushConditionIntoOrShuttle extends RexShuttle {
+
+    private RexBuilder rexBuilder;
+    private RexSimplify simplify;
+
+    public PushConditionIntoOrShuttle(RexSimplify simplify) {
+      super();
+      this.rexBuilder = simplify.rexBuilder;
+      this.simplify = simplify;
+    }
+
+    @Override public RexNode visitCall(RexCall call) {
+      for (;;) {
+        call = (RexCall) super.visitCall(call);
+        final RexCall old = call;
+        call = processCall(call);
+        if (call == old) {
+          return call;
+        }
+      }
+    }
+
+    /** Helper class to represent compareOps / disjunctive forms which are connected to
+     * the same variable */
+    class CandidateGroup {
+
+      List<RexNode> compareOp = new ArrayList<>();
+      List<RexNode> dfs = new ArrayList<>();
+      private RexNode result = null;
+
+      public void addMember(RexNode rexNode) {
+        if (rexNode.getKind() == SqlKind.OR) {
+          dfs.add(rexNode);
+        } else {
+          compareOp.add(rexNode);
+        }
+      }
+
+      public boolean pushIn() {
+        if (dfs.size() != 1 || compareOp.size() != 1) {
+          return false;
+        }
+        // in case this pushIn is not successfull; someone might pull out the "AND-ed operands"
+        boolean reduced = false;
+        List<RexNode> orOps = ((RexCall) dfs.get(0)).getOperands();
+        RexCall cmpOp = (RexCall) compareOp.get(0);
+        List<RexNode> newOps = new ArrayList<>();
+        for (RexNode rexNode : orOps) {
+          RexNode newCall = rexBuilder.makeCall(SqlStdOperatorTable.AND, rexNode, cmpOp);
+          newCall = simplify.simplify(newCall);
+          if (newCall.isAlwaysFalse() || newCall.isAlwaysTrue()) {
+            // we have successfully burried an operand
+            reduced = true;
+          }
+          newOps.add(newCall);
+        }
+        result = rexBuilder.makeCall(SqlStdOperatorTable.OR, newOps);
+        result = simplify.simplify(result);
+        if (result.getKind() != SqlKind.OR) {
+          reduced = true;
+        }
+        return reduced;
+      }
+
+      public List<RexNode> getResults() {
+        if (result != null) {
+          return Lists.newArrayList(result);
+        }
+        List<RexNode> ret = new ArrayList<>();
+        ret.addAll(compareOp);
+        ret.addAll(dfs);
+        return ret;
+      }
+
+    }
+
+    private RexCall processCall(RexCall call) {
+      List<RexNode> others = new ArrayList<>();
+
+      if (call.getKind() != SqlKind.AND) {
+        return call;
+      }
+
+      List<RexNode> ops = call.getOperands();
+      Map<ImmutableBitSet, CandidateGroup> candidateMap = new HashMap<>();
+
+      for (RexNode rexNode : ops) {
+        ImmutableBitSet k = RelOptUtil.InputFinder.bits(rexNode);
```

Comment:
Does this look like a O(N^2)?

`visitCall` visits every node in the tree, and `RelOptUtil.InputFinder.bits` processes the whole subtree.

I wonder if the overhead is worth the optimization produced.

Reply:
sorry @vlsi  this is an old PR; this approach was abandoned.....
I will now look around and close any pr which seem outdated.